homeassistant/components/imap/config_flow.py

```diff
@@ -0,0 +1,123 @@
+"""Config flow for imap integration."""
+from __future__ import annotations
+
+from asyncio import TimeoutError as AsyncIOTimeoutError
+from collections.abc import Mapping
+from typing import Any
+
+from aioimaplib import AioImapException
+import voluptuous as vol
+
+from homeassistant import config_entries
+from homeassistant.const import CONF_NAME, CONF_PASSWORD, CONF_PORT, CONF_USERNAME
+from homeassistant.data_entry_flow import FlowResult
+from homeassistant.helpers import config_validation as cv
+
+from .const import (
+    CONF_CHARSET,
+    CONF_FOLDER,
+    CONF_SEARCH,
+    CONF_SERVER,
+    DEFAULT_PORT,
+    DOMAIN,
+)
+from .coordinator import connect_to_server
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_USERNAME): str,
+        vol.Required(CONF_PASSWORD): str,
+        vol.Required(CONF_SERVER): str,
+        vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+        vol.Optional(CONF_CHARSET, default="utf-8"): str,
+        vol.Optional(CONF_FOLDER, default="INBOX"): str,
+        vol.Optional(CONF_SEARCH, default="UnSeen UnDeleted"): str,
```

Comment:
Can we move charset, folder and search to an options flow, or could the connection fail if they are incorrectly set? We only want the required items for the connection in the main flow.

Reply:
- For 'charset' this doesn't change for a configured mailbox and choosing the wrong charset will return a bad result response when doing the initial update. I recommend to keep it as part of the main flow.
- As for the folder and search fields, we can move them to options flow. My concern is that there is a chance a user configures multiple instances with the same mailbox (username) and selects the same folder and search values (ending up with duplicate sensors). If that is acceptable I can surely move them to options.

Reply:
Yeah, I guess it's more convenient to keep all items in the main flow. Otherwise the user will need to set up multiple config entries with different searches or inboxes in a specific order, ie the non default once first.